### PR TITLE
8319406: x86: Shorter movptr(reg, imm) for 32-bit immediates

### PR DESCRIPTION
**Pull request by shipilev**
Noticed this while doing C1 work, but the issue is more generic. If you look into x86 code, then sometimes you'll notice `movabs` with small immediates on x86. That `movabs` actually carries the full-blown 64-bit immediate.

Similar to [JDK-8255838](https://bugs.openjdk.org/browse/JDK-8255838), it would be useful to shorten movptr(reg, imm) when immediate fits in 32 bits. This would compact some code, notably the code in C1 profiling ([JDK-8315843](https://bugs.openjdk.org/browse/JDK-8315843)), but also other code, generically.

For example, sample branch profiling hunk from C1 tier3 on x86_64:

```
Before:
   0x00007f269065ed02:   test   %edx,%edx
   0x00007f269065ed04:   movabs $0x7f260a4ddd68,%rax  ;   {metadata(method data for {method} …
   0x00007f269065ed0e:   movabs $0x138,%rsi
 ╭ 0x00007f269065ed18:   je     0x00007f269065ed24
 │ 0x00007f269065ed1a:   movabs $0x148,%rsi
 ↘ 0x00007f269065ed24:   mov    (%rax,%rsi,1),%rdi
   0x00007f269065ed28:   lea    0x1(%rdi),%rdi
   0x00007f269065ed2c:   mov    %rdi,(%rax,%rsi,1)
   0x00007f269065ed30:   je     0x00007f269065ed4e     

After:
   0x00007f1370dcd782:   test   %edx,%edx
   0x00007f1370dcd784:   movabs $0x7f12f64ddd68,%rax   ;   {metadata(method data for {method} …
   0x00007f1370dcd78e:   mov    $0x138,%esi
 ╭ 0x00007f1370dcd793:   je     0x00007f1370dcd79a        
 │ 0x00007f1370dcd795:   mov    $0x148,%esi
 ↘ 0x00007f1370dcd79a:   mov    (%rax,%rsi,1),%rdi        
   0x00007f1370dcd79e:   lea    0x1(%rdi),%rdi
   0x00007f1370dcd7a2:   mov    %rdi,(%rax,%rsi,1)       
   0x00007f1370dcd7a6:   je     0x00007f1370dcd7c4 
```

We can use shorter 32-bit immediate moves. In the hunk above, this saves about 8 bytes, look around `movabs` -> `mov` changes. But this is not limited to the profiling code. There are nearly 1% code space savings on larger tests in C2. For example, on `-Xcomp -XX:TieredStopAtLevel=... HelloWorld`:

```
# Before
 nmethod code size         :   430328 bytes
 nmethod code size         :   467032 bytes
 nmethod code size         :   908936 bytes
 nmethod code size         :  1267816 bytes

# After
 nmethod code size         :   429616 bytes (-0.1%)
 nmethod code size         :   466344 bytes (-0.1%)
 nmethod code size         :   897144 bytes (-1.3%)
 nmethod code size         :  1256216 bytes (-0.9%)
```

There are two wrinkles:
  1. Current `movslq(Register, int32_t)` is broken and protected by `ShouldNotReachHere()`. I would have used it in this patch, but x86_64 does not actually define `movslq reg64, imm32`, so we use a regular `mov reg64, imm32`. Removed the broken definition too.
  2. There is at least one place in Hotspot -- IC calls -- that expects the synthetic `movptr` to always have the same length, because it would be used as IC slot. I looked through other uses of `movptr(Register, intptr_t)`, and no other are suspicious. 

Additional testing:
 - [x] Linux x86_64 server fastdebug, `tier1 tier2 tier3 tier4`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319406](https://bugs.openjdk.org/browse/JDK-8319406): x86: Shorter movptr(reg, imm) for 32-bit immediates (**Enhancement** - P4)

### Reviewers
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) ⚠️ Review applies to [6dcaf425](https://git.openjdk.org/jdk/pull/16497/files/6dcaf425fbfcc090e1ab4dcfac67a24950985234)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16497/head:pull/16497` \
`$ git checkout pull/16497`

Update a local copy of the PR: \
`$ git checkout pull/16497` \
`$ git pull https://git.openjdk.org/jdk.git pull/16497/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16497`

View PR using the GUI difftool: \
`$ git pr show -t 16497`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16497.diff">https://git.openjdk.org/jdk/pull/16497.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16497#issuecomment-1792947974)

### CODE REVIEW

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 1]
```diff
@@ -2562,7 +2571,15 @@ void MacroAssembler::movptr(Register dst, Address src) {
 
 // src should NEVER be a real pointer. Use AddressLiteral for true pointers
 void MacroAssembler::movptr(Register dst, intptr_t src) {
-  LP64_ONLY(mov64(dst, src)) NOT_LP64(movl(dst, src));
+#ifdef _LP64
+  if (is_simm32(src)) {
+    movslq(dst, checked_cast<int32_t>(src));
```
Comment (merykitty) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:2576]: Why not just `movq`? there is no `movslq r, i` so this is kind of confusing.

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 2]
```diff
@@ -1815,6 +1815,8 @@ class MacroAssembler: public Assembler {
   void mov_metadata(Register dst, Metadata* obj);
   void mov_metadata(Address  dst, Metadata* obj, Register rscratch);
 
+  void mov_ptrslot(Register dst, intptr_t val);
```
Comment (merykitty) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:1818]: I believe the convention here would be `movptr_imm64`

**Review by merykitty — COMMENTED** [review]
Can we create `MacroAssembler::mov64` that does the branching instead, I think it is more natural there. And things that need 8-byte immediates will call into `Assembler::mov64`.

Thanks.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:2576]: Right. There is no point in trying to fix `movslq`. Replaced with `movq`.

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/cpu/x86/macroAssembler_x86.hpp:1818]: Good name, renamed.

**Comment by shipilev** [comment]
> Can we create `MacroAssembler::mov64` that does the branching instead, I think it is more natural there. And things that need 8-byte immediates will call into `Assembler::mov64`.

Right, that would capture more cases. I am a bit concerned that we would not catch all the places where `mov64` is actually needed to be full-blown immediate. Maybe it would be safer to keep `mov64` intact, and introduce `mov_immediate` that can be arbitrarily shortened? AArch64 does this, for example.

**Comment by shipilev** [comment]
I simplified the patch a bit, still looking for reviewers. @merykitty, does the new version look better to you?

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 3]
```diff
@@ -13450,17 +13450,6 @@ void Assembler::movsbq(Register dst, Register src) {
   emit_int24(0x0F, (unsigned char)0xBE, (0xC0 | encode));
 }
 
-void Assembler::movslq(Register dst, int32_t imm32) {
```
Comment (merykitty) [inline src/hotspot/cpu/x86/assembler_x86.cpp:13453]: You can remove the corresponding declaration in the header file.

**Review by merykitty — APPROVED** [review]
LGTM

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 3 above)
Comment (shipilev) [reply src/hotspot/cpu/x86/assembler_x86.cpp:13453]: Right, removed.

**Review by vnkozlov — APPROVED** [review]
Looks good to me too.

**Comment by shipilev** [comment]
Thanks! Testing passes, so I am integrating.

/integrate

**Pull request closed without merging**